shell/platform/windows/compositor_software.cc

```diff
@@ -47,18 +119,46 @@ bool CompositorSoftware::Present(FlutterWindowsView* view,
     return view->ClearSoftwareBitmap();
   }
 
-  // TODO: Support compositing layers and platform views.
-  // See: https://github.com/flutter/flutter/issues/31713
-  FML_DCHECK(layers_count == 1);
-  FML_DCHECK(layers[0]->offset.x == 0 && layers[0]->offset.y == 0);
-  FML_DCHECK(layers[0]->type == kFlutterLayerContentTypeBackingStore);
-  FML_DCHECK(layers[0]->backing_store->type ==
-             kFlutterBackingStoreTypeSoftware);
+  // Bypass composition logic if there is only one layer.
+  if (layers_count == 1) {
+    const FlutterLayer* layer = layers[0];
+    FML_DCHECK(layer != nullptr);
+    if (layer->type == kFlutterLayerContentTypeBackingStore &&
+        layer->offset.x == 0 && layer->offset.y == 0) {
+      auto& backing_store = *layer->backing_store;
+      FML_DCHECK(backing_store.type == kFlutterBackingStoreTypeSoftware);
+      auto& software = backing_store.software;
+      return view->PresentSoftwareBitmap(software.allocation,
+                                         software.row_bytes, software.height);
+    }
+  }
+
+  // Composite many layers.
+  FlutterRect bounds = CalculateBounds(layers, layers_count);
+  // Truncate from double to integer to represent whole pixels.
+  int x_min = static_cast<int>(bounds.left);
+  int x_max = static_cast<int>(bounds.right);
+  int y_min = static_cast<int>(bounds.top);
+  int y_max = static_cast<int>(bounds.bottom);
 
-  const auto& backing_store = layers[0]->backing_store->software;
+  int width = x_max - x_min;
+  int height = y_max - y_min;
+  std::vector<uint32_t> allocation(width * height, kOpaqueBlack);
+
+  for (const FlutterLayer** layer = layers; layer < layers + layers_count;
+       layer++) {
+    // TODO(schectman): handle platform view type layers.
+    // https://github.com/flutter/flutter/issues/143375
+    if ((*layer)->type == kFlutterLayerContentTypeBackingStore) {
+      BlendLayer(allocation, **layer, x_min, y_min, width, height);
+    } else {
+      FML_UNREACHABLE();
+      return false;
+    }
+  }
 
-  return view->PresentSoftwareBitmap(
-      backing_store.allocation, backing_store.row_bytes, backing_store.height);
+  return view->PresentSoftwareBitmap(static_cast<void*>(allocation.data()),
+                                     width * 4, height);
```

Comment:
For someone not familiar with the bitmap allocation, this `4` might look like a magic value. Could we add a constant for the 4 with a quick comment for clarity?